### PR TITLE
API: bump MAXDIMS/MAXARGS to 64 introduce NPY_AXIS_RAVEL

### PR DESCRIPTION
**Pull request by seberg**
Bumping these two has three main caveats:
1. We cannot break ABI for the iterator macros, which means the
   iterators now need to use 32/their own macro.
2. We used axis=MAXDIMS to mean axis=None, introduce NPY_AXIS_RAVEL
   to replace this, it will be run-time outside NumPy.
3. The old style iterators cannot deal with high dimensions, meaning
   that some functions just won't work (because they use them).

---

**Draft, as based on the mapiter removal**.  ~Needs to document the above points, but~ should be good for an initial review ~(the doc changes here may be a bit random).  But,~ I don't expect the code to change really.

There are two plausible follow-ups that I would prefer _not_ to include here:
1. I had some start to use `alloca()` or similar for some allocations (or malloc for big ones).  We don't have to do go all the way there, but at least the stack arrays in the ufunc setup are worth shrinking.  (definitely want to do that)
2. It might be nice to address some of the places that remain limited.  Examples are the `choose` and `arr.flat` I added test for here.  As well as `np.random` since it relies on the old-style iterator a lot.
   * I don't have a real solution here, we could duplicate the old style iterators, giving you more dimensions (if you rely on NumPy >2, backporting isn't plausible).

---

I had mentioned before to just delete `NPY_MAXDIMS` as a public symbol rather than deleting it.  After seeing that it is used in a few places in SciPy, etc. I decided that maybe it is best to defer that to NumPy 3.  I doubt we are ready to force that churn on downstream.

Addresses gh-5744, although some paths remain of course.

### CODE REVIEW

**Comment by seberg** [comment]
OK, I think this is now ready (but relying on the other PR).  I had not done it here, but I would be happy to follow up with:
* Removing `NPY_MAXARGS`.  The main useful "public" use seems at the end of the iterator structs (and it is OK to change it there).  (could also do this here, but not sure it helps)
* I checked and Cython does not define `arr.flat` to return a C iterator.  So I think we can make `arr.flat` return a new object that supports more dimensions.
  This may not be a super quick thing, unfortunately.  Unless we mostly copy paste the old one.  The reason is that I am not sure the indexing code can be redone very easily.  (Even if it would be nice anyway)

**Comment by mattip** [comment]
> the other PR

Link?

**Comment by seberg** [comment]
Oops, thought it was in the first post, just opened: https://github.com/numpy/numpy/pull/25138

File: numpy/_core/src/multiarray/conversion_utils.c [hunk 1]
```diff
@@ -325,23 +325,14 @@ NPY_NO_EXPORT int
...
         }
-        if (*axis == NPY_MAXDIMS){
-            /* NumPy 1.23, 2022-05-19 */
-            if (DEPRECATE("Using `axis=32` (MAXDIMS) is deprecated. "
-                          "32/MAXDIMS had the same meaning as `axis=None` which "
-                          "should be used instead.  "
-                          "(Deprecated NumPy 1.23)") < 0) {
-                return NPY_FAIL;
-            }
-        }
```
Comment (seberg) [inline numpy/_core/src/multiarray/conversion_utils.c:344]: This is very niche, but, mentioned it in the release notes.

File: numpy/_core/src/multiarray/multiarraymodule.c [hunk 2]
```diff
@@ -692,7 +692,7 @@ PyArray_ConcatenateInto(PyObject *op,
         Py_DECREF(item);
     }
 
-    if (axis >= NPY_MAXDIMS) {
+    if (axis == NPY_RAVEL_AXIS) {
```
Comment (seberg) [inline numpy/_core/src/multiarray/multiarraymodule.c:695]: Also a subtle oddity, mentioned in the release note (any huge axes was interpreted as raveling)

File: numpy/_core/src/multiarray/refcount.c [hunk 3]
```diff
@@ -340,6 +347,13 @@ PyArray_XDECREF(PyArrayObject *mp)
         }
     }
     else { /* handles misaligned data too */
+        if (PyArray_NDIM(mp) > NPY_MAXDIMS_LEGACY_ITERS) {
+            PyErr_Format(PyExc_RuntimeError,
+                    "this function only supports up to 32 dimensions but "
+                    "the array has %d.", PyArray_NDIM(mp));
+            return -1;
+        }
```
Comment (seberg) [inline numpy/_core/src/multiarray/refcount.c:355]: We don't us this internally, might remove the function, but not sure yet.  In principle, anyone running into it should catch the error and not leak anything (normal array cleanup doesn't have this problem).

File: numpy/__init__.pxd [hunk 4]
```diff
@@ -182,9 +182,8 @@ cdef extern from "numpy/arrayobject.h":
         NPY_ARRAY_UPDATE_ALL
 
     cdef enum:
-        NPY_MAXDIMS
-
-    npy_intp NPY_MAX_ELSIZE
```
Comment (seberg) [inline numpy/__init__.pxd:187]: Maybe a bit random, but `NPY_MAX_ELSIZE` doesn't exist (it sure did at some point).

**Comment by seberg** [comment]
@ngoldbaum maybe you have some time to review this?

File: doc/source/numpy_2_0_migration_guide.rst [hunk 5]
```diff
@@ -41,8 +41,30 @@ Some are defined in ``numpy/_core/include/numpy/npy_2_compat.h``
...
 Please let us know if you require additional workarounds here.
 
+.. _migration_maxdims:
+
+Increased maximum number of dimensions
+--------------------------------------
+The maximum number of dimensions (and arguments) was increased to 64, this
+affects the ``NPY_MAXDIMS`` and ``NPY_MAXARGS`` macros.
+It may be good to review their use, and we generally encourage you to
+not use this macros (especially ``NPY_MAXARGS``), so that a future version of
```
Comment (ngoldbaum) [inline doc/source/numpy_2_0_migration_guide.rst:57]:
```suggestion
not use these macros (especially ``NPY_MAXARGS``), so that a future version of
```

File: doc/release/upcoming_changes/25149.c_api.rst [hunk 6]
```diff
@@ -0,0 +1,10 @@
+Lager ``NPY_MAXDIMS`` and ``NPY_MAXARGS``, ``NPY_RAVEL_AXIS`` introduced
+------------------------------------------------------------------------
+
+``NPY_MAXDIMS`` is now 64, you may want to review its use.  This is usually
+used in stack allocation, where the increase should be safe.
+However, we do encourage generally to remove any use of ``NPY_MAXDIMS`` and
+``NPY_MAXARGS`` to eventually allow removing the constraint completely.
+In some cases, ``NPY_MAXDIMS`` was passed (and returned) to mean ``axis=None``
```
Comment (ngoldbaum) [inline doc/release/upcoming_changes/25149.c_api.rst:8]: Can you make "In some cases" a little more specific? Does this only impact things that interact with numpy via the C API, or could you trigger this behavior from the python API too?

File: doc/release/upcoming_changes/25149.c_api.rst [hunk 7]
```diff
@@ -0,0 +1,10 @@
+Lager ``NPY_MAXDIMS`` and ``NPY_MAXARGS``, ``NPY_RAVEL_AXIS`` introduced
+------------------------------------------------------------------------
+
+``NPY_MAXDIMS`` is now 64, you may want to review its use.  This is usually
+used in stack allocation, where the increase should be safe.
```
Comment (ngoldbaum) [inline doc/release/upcoming_changes/25149.c_api.rst:5]:
```suggestion
used in a stack allocation, where the increase should be safe.
```

File: doc/release/upcoming_changes/25149.change.rst [hunk 8]
```diff
@@ -0,0 +1,8 @@
+Out-of-bound axis not the same as ``axis=None``
+-----------------------------------------------
+In some cases ``axis=32`` or for concatenate any large value
+was the same as ``axis=None``.
+Except for ``concatenate`` this was deprecate.
```
Comment (ngoldbaum) [inline doc/release/upcoming_changes/25149.change.rst:5]:
```suggestion
Except for ``concatenate`` this was deprecated in Numpy 1.23.
```

Was the deprecation noisy for other functions? Just wondering if we might want to merely deprecate this usage for `concatenate` since that was missed in 2022.

File: doc/source/numpy_2_0_migration_guide.rst [hunk 9]
```diff
@@ -41,8 +41,30 @@ Some are defined in ``numpy/_core/include/numpy/npy_2_compat.h``
...
+NumPy can remove this limitation on the number of dimensions.
+
+``NPY_MAXDIMS`` was also used to signal ``axis=None`` in the C-API, including
+the ``PyArray_AxisConverter``.
+If you run into this problem, you will see ``-2147483648``
+(the minimum integer value) or ``64`` being used as an invalid axis.
```
Comment (ngoldbaum) [inline doc/source/numpy_2_0_migration_guide.rst:63]: It might help to include the exact text of the error they'd see, so if they google the error message they'll find the migration guide

File: doc/source/reference/c-api/array.rst [hunk 10]
```diff
@@ -3411,11 +3411,26 @@ Other constants
...
 
-    The maximum number of dimensions allowed in arrays.
+    The maximum number of dimensions that may be used by NumPy.
+    This is set to 64 and was 32 before NumPy 2.
+
+    .. note::
+        We encourage you to avoid ``NPY_MAXDIMS``.  A future version of NumPy
+        may wish to remove any dimension limitation (and thus the constant).
+        The limitation was created mainly a simplification for the liberal use
+        of small stack allocations as scratch space.
```
Comment (ngoldbaum) [inline doc/source/reference/c-api/array.rst:3421]:
```suggestion
        The limitation was created so that NumPy can use stack allocations
        internally for scratch space.
```

File: doc/source/reference/c-api/array.rst [hunk 11]
```diff
@@ -3411,11 +3411,26 @@ Other constants
...
+
+        If your algorithm has a reasonable maximum number of dimension you
+        could check and use that locally.
 
 .. c:macro:: NPY_MAXARGS
 
-    The maximum number of array arguments that can be used in functions.
+    The maximum number of array arguments that can be used in in some
```
Comment (ngoldbaum) [inline doc/source/reference/c-api/array.rst:3428]:
```suggestion
    The maximum number of array arguments that can be used in some
```

File: doc/source/reference/c-api/array.rst [hunk 12]
```diff
@@ -3435,6 +3450,14 @@ Other constants
     The return value of successful converter functions which are called
     using the "O&" syntax in :c:func:`PyArg_ParseTuple`-like functions.
 
+.. c:macro:: NPY_RAVEL_AXIS
+
+    Some NumPy functions (mainly the C-entrypoints for Python functions)
+    have an ``axis`` argument.  This macro may be passed for ``axis=None``.
+
+    .. note::
+        This macro is a NumPy version dependent at runtime.
```
Comment (ngoldbaum) [inline doc/source/reference/c-api/array.rst:3459]:
```suggestion
        This macro is NumPy version dependent at runtime.
```

Maybe also a sentence explaining why it depends on Numpy version and/or why that matters?

File: doc/source/reference/c-api/types-and-structures.rst [hunk 13]
```diff
@@ -119,8 +119,11 @@ PyArray_Type and PyArrayObject
        array. When nd is 0, the array is sometimes called a rank-0
        array. Such arrays have undefined dimensions and strides and
        cannot be accessed. Macro :c:data:`PyArray_NDIM` defined in
-       ``ndarraytypes.h`` points to this data member. :c:data:`NPY_MAXDIMS`
-       is the largest number of dimensions for any array.
+       ``ndarraytypes.h`` points to this data member.
+       Although most operations may be limited in dimensionality, we do not
+       advertise a maximum dimension.  Anyone explicitly relying on one
```
Comment (ngoldbaum) [inline doc/source/reference/c-api/types-and-structures.rst:124]: Is this a remnant of an earlier version that completely removed `NPY_MAXDIMS` from the public API? Maybe rephrase this to say `NPY_MAXDIMS` is there but not to rely on it because it might change or be removed in the future.

File: numpy/__init__.pxd (hunk 4 above)
Comment (ngoldbaum) [reply numpy/__init__.pxd:187]: Does this removal need a release note? Just to say that if someone was using and relying on it, it wasn't doing what they thought it did.

**Review by ngoldbaum — COMMENTED** [review]
A few language suggestions and some questions. Overall looks fine though.

**Review by ngoldbaum — APPROVED** [review]
I'm on vacation but will merge this next week if no one else does it before me. This is mostly very straightforward and it's hard to know the impact of the ABI break or possible breakage of questionable C code using the macro without making the change.

**Comment by ngoldbaum** [comment]
I'm going to merge this now and immediately trigger a new nightly wheel upload. Thanks @seberg!

If you are reading this because of new CI breakage, make sure any nightly builds or tests you're running against NumPy dev (future 2.0 release) use a stack that is built against Numpy dev as well, with build isolation disabled.
